modules/fedimint-ln-client/src/lib.rs

```diff
@@ -319,42 +336,50 @@ impl LightningClientExt for Client {
 
         let refund_success = lightning.await_refund(operation_id);
 
-        Ok(Box::pin(stream! {
+        let stream = stream! {
             yield LnPayState::Created;
 
-            match tx_accepted_future.await {
-                Ok(()) => {
-                    yield LnPayState::Funded;
+            if tx_accepted_future.await.is_err() {
+                yield LnPayState::Canceled;
+                return;
+            }
 
-                    // await success or waiting for refund
-                    match payment_success.await {
-                        Ok(preimage) => {
-                            yield LnPayState::Success {preimage};
-                        }
-                        Err(LightningPayError::Refundable(refundable)) => {
-                            yield LnPayState::WaitingForRefund{ block_height: refundable };
-
-                            // in case of refund, await output of refund
-                            match refund_success.await {
-                                Ok(refund_txid) => {
-                                    yield LnPayState::Refunded { refund_txid };
-                                }
-                                Err(_) => {
-                                    yield LnPayState::Failed;
-                                }
+            yield LnPayState::Funded;
+
+            match payment_success.await {
+                Ok(preimage) => {
+                    if let Some(change) = change_outpoint {
+                        yield LnPayState::AwaitingChange;
+                        match self.await_primary_module_output_finalized(operation_id, change).await {
+                            Ok(_) => {}
+                            Err(_) => {
+                                yield LnPayState::Failed;
```

Comment:
There is a nuance here that a malicious federation could block the issuance of e-cash while the contract is funded and the payment succeeds, but probably not a priority to fix.